src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -283,6 +285,89 @@ public class LockFileBuilder
             return lockFile;
         }
 
+        private void ValidateLockFileLibrary(LockFile lockFile)
+        {
+            IList<LockFileLibrary> libraries = lockFile.Libraries;
+            var libraryReferences = new Dictionary<string, LockFileLibrary>();
+
+            foreach (LockFileLibrary library in libraries)
+            {
+                var libraryKey = library.Name + " " + library.Version;
+
+                if (libraryReferences.TryGetValue(libraryKey, out LockFileLibrary existingLibrary))
+                {
+                    if (RankReferences(existingLibrary.Type) > RankReferences(library.Type))
+                    {
+                        // Prefer project reference over package reference, so replace the the package reference.
+                        libraryReferences[libraryKey] = library;
+                    }
+                }
+                else
+                {
+                    libraryReferences[libraryKey] = library;
+                }
+            }
+
+            lockFile.Libraries = new List<LockFileLibrary>(libraryReferences.Count);
```

Comment:
Would it make sense to *only* do this when the libraryReferences count is *not* equivalent to the libraries count? 

That means we did something. No need to allocate beyond that.

Reply:
Sure, thank you for your review. 
I was doing it in initial implementation differently, but missed in new implementation. 
Please check now.